### ISSUE TITLE
Changes keep jumping from my branch to the default branch

### ISSUE THREAD

**Issue by TheYarin** [gitbutlerapp/gitbutler#4960]
### Version

0.12.25

### Operating System

Windows

### Distribution Method

msi (Windows)

### Describe the issue

I constantly encounter the following bug: I drag changes from the default branch to another branch, and after a few seconds the changes jump back to the default branch. I drag them again, they move, and then after a few more seconds they jump back. I find myself having to quickly commit them before they jump back.

### How to reproduce

Not sure, this happens often though

### Expected behavior

well, I expect the changes to stay in the branch I put them in.

**Comment by Byron** [comment]
Thanks a lot for reporting!

It's sad to see the issue is so spurious. From a quick attempt I can say that it doesn't obviously reproduce - files dragged over to other virtual branches stay there, even when the app is refreshed either by hand or because the worktree changes.

Maybe you can find out what affects the issue, as it doesn't seem to always occur.

**Comment by TheYarin** [comment]
hmm, that sounds not so straight forward to pinpoint. is there any piece of code that might move changes from a virtual branch back to the default one? or maybe an operation that "resets" the state of GitButler?

I did notice a wider problem that takes form as many symptoms and might be related here, too: once an operation (such as moving changes between branches) starts, it captures (at least) the state of its original branch and if the operation takes a while and other operations are started in the meantime, they will effectively override any previous operations. Maybe this issue manifests here as well?

**Comment by Byron** [comment]
Yes, the app is full of race conditions. However, by now we do queue them but maybe the function causing this isn't actually using this system. Typically I only added the system (i.e. a mutex) where it was obvious.

CC @Caleb-T-Owens , as he might now better which mutating functions in the application aren't using the exclusive access lock, making them racy.

**Comment by Caleb-T-Owens** [comment]
@Byron the `move_commit_file` is the one in question I believe. That said, most of the functions in `virtual.rs` ought to have the lock passed into them as well.

**Comment by Caleb-T-Owens** [comment]
@TheYarin Thanks for writing in.

I'll see if I can find some time this week to shore up our worktree write locks because that is unacceptable.

**Comment by Byron** [comment]
Thanks a lot @Caleb-T-Owens for chiming in!

This function definitely doesn't grab the lock to serialize it, which explains its raciness.

https://github.com/gitbutlerapp/gitbutler/blob/7dbadb13a9d36e31bf272a7964fb4573f958c9e6/crates/gitbutler-branch-actions/src/actions.rs#L291-L292

It will be quite some work to propagate it to all relevant places, but in return I'd expect all issues like these fixed.

On another note, it's interesting how serial the app has become, and that this is a good thing as that makes it deterministic as well. Taking this upon ourselves instead of just using non-async tauri commands seems beneficial overall as it helps differentiate mutating functions from read-only ones in an application that doesn't benefit from Rusts usual rules around mutation as all state lives on disk.

**Comment by TheYarin** [comment]
Glad to see you guys narrowed it down!

On another note, I had some suspicion this issue might be triggered by my IDE so I checked and can confirm this happens even when the IDE is closed.

**Comment by krlvi** [comment]
Hey @TheYarin, sorry this sat around for so long. The branch-assignment system from 0.12.25 has since been replaced with a persisted hunk-assignment system in https://github.com/gitbutlerapp/gitbutler/pull/8612, and the new UI became the default in release 0.15.0: https://github.com/gitbutlerapp/gitbutler/releases/tag/release/0.15.0

Since the original report is tied to the old system and we no longer have a reproduction for it, im going to close this one. If changes still become unassigned on the latest release, please open a fresh issue with the exact action that triggers it. The specific case where leaving Edit mode clears assignments is already tracked in #10987.
